girder/utility/abstract_assetstore_adapter.py

```diff
@@ -25,6 +26,87 @@
 from .model_importer import ModelImporter
 
 
+class FileHandle(object):
+    """
+    This is the base class that is returned for the file-like API into
+    Girder file objects. The ``open`` method of assetstore implementations
+    is responsible for returning an instance of this class or one of its
+    subclasses. This base class implementation is returned by the
+    abstract assetstore adapter, and does not leverage any details of the
+    assetstore implementations.
+
+    These file handles are stateful, and therefore not safe for concurrent
+    access. If used by multiple threads, mutexes should be used.
+
+    :param file: The file object to which this file-like object corresponds.
+    :type file: dict
+    :param adapter: The assetstore adapter corresponding to this file.
+    :type adapter: girder.utility.abstract_assetstore_adapter.AbstractAssetstoreAdapter
+    """
+    def __init__(self, file, adapter):
+        self._file = file
+        self._adapter = adapter
+        self.seek(0)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def read(self, size):
+        """
+        Read *size* bytes from the file data.
+
+        :param size: The number of bytes to read from the current position. The
+            actual number returned could be less than this if the end of the
+            file is reached. An empty response indicates that the file has been
+            completely consumed.
+        :type size: int
+        :rtype: bytes
+        """
+        data = six.BytesIO()
+        length = 0
+        for chunk in itertools.chain(self._prev, self._stream):
+            chunkLen = len(chunk)
+
+            if chunkLen == 0:
+                break
+
+            if length + chunkLen <= size:
+                data.write(chunk)
+                self._prev = []
+                length += chunkLen
+
+                if length + chunkLen == size:
+                    break
+            else:
+                chunkLen = min(size - length, chunkLen)
+                data.write(chunk[:chunkLen])
+                self._prev = [chunk[chunkLen:]]
+                length += chunkLen
+                break
+
+        self._pos += length
+        return data.getvalue()
+
+    def tell(self):
+        return self._pos
+
+    def seek(self, offset, whence=os.SEEK_SET):
+        if whence == os.SEEK_SET:
+            self._pos = offset
+        elif whence == os.SEEK_CUR:
+            self._pos += offset
+        elif whence == os.SEEK_END:
+            self._pos = self._file['size'] - offset
+        self._prev = []
+        self._stream = self._adapter.downloadFile(self._file, offset=self._pos, headers=False)()
```

Comment:
Because creating a new stream could have significant overhead, for seeks that don't change the file position, I think it would be better to have a check that if `self._pos` doesn't change then we don't recreate the `self._stream` value.